protocols/kad/src/behaviour.rs

```diff
@@ -990,6 +994,40 @@ where
         id
     }
 
+    pub fn set_explicit_mode(&mut self, mode: Option<Mode>) {
+        match mode {
+            Some(mode) => {
+                self.mode = mode;
+                self.auto_mode = false;
+                if !self.connections.is_empty() {
+                    self.reconfigure_mode();
+                }
+            }
+            None => {
+                self.auto_mode = true;
+            }
```

Comment:
Say the user has previously explicitly set the mode to `Mode::Client`. Now the user calls `set_explicit_mode(None)`, thus going into auto-mode, and thus this match arm is executed.

In case we have an external address, we would now be in `Mode::Server`. In such case all `ConnectionHandler`s need to be notified. But as far as I can tell this is not the case, i.e. `self.mode` is not re-determined (like it is done in `on_swarm_event`) and the handlers are not updated (like it is done in `reconfigure_mode`.

Would it make sense to pull the mode related logic out of `on_swarm_event` and combine it into a single method with `reconfigure_mode`. That method would then be called in two cases (1) `on_swarm_event` with a new external address and (2) when the user calls `set_explicit_mode`?

Reply:
I do agree on that. I guess it never got pushed up, but I did made some changes to move that logic into a single function that is called from either `set_explicit_mode` or `on_swarm_event` when `auto_mode` is true so it could determined the mode based on the external addresses since I wasnt really sure if I wanted have it to depend on just `on_swarm_event` to make determine that change and to notify the handlers when it could be done in `set_explicit_mode`.